### PR TITLE
feat: add interactive /claude:plugin browse picker

### PR DESCRIPTION
**Pull request by rakesh-vs**
## Summary

Adds an interactive `SelectList` browser for `/claude:plugin`, mirroring Claude Code's `/plugin` UX.

- `/claude:plugin browse`: explicit browser invocation
- bare `/claude:plugin`: opens the browser in interactive (TUI) mode, falls back to `list` in non-TUI mode

Four screens: **marketplaces -> plugins -> actions -> install scope (for install)**. Esc walks back; Enter descends or commits. The chosen action dispatches to the existing typed subcommand handlers (`install`/`uninstall`/`info`/`enable`/`disable`) with a constructed `<plugin>@<marketplace> --scope <scope> [--local]` arg.

## What's added

- `edge/browser/plugin-browser.ts`: `PluginBrowser` component. Pure UI: data loading is injected via `pluginLoader`, with no `ctx`/`pi`/orchestrator coupling. Built on `SelectList` + `DynamicBorder`.
  - Includes a dedicated **Scope Selection** screen on install offering:
    1. `Project local (Recommended)`: writes to `.pi/claude-plugins.local.json` (`--local`, gitignored)
    2. `Project`: writes to `.pi/claude-plugins.json` (committed, repo-shared)
    3. `User (global)`: writes to `~/.pi/agent/claude-plugins.json` (all projects)
- `edge/handlers/plugin/browse.ts`: handler with a TUI guard (non-TUI modes fall back to `list`) + the real `pluginLoader` wrapping `loadPluginListPayload`.
- `platform/pi-api.ts`: chokepoint re-exports of `DynamicBorder` (runtime) and `Theme` (type), keeping BLOCK E intact.
- Router/register wiring + `browse` in autocomplete.
- Direct test coverage: 100% line, branch, and function coverage across `plugin-browser.ts` (15 tests) and `browse.ts` (12 tests).
- Standalone demo: `demos/browse-demo.ts` + `demos/run-browse-demo.sh`.
- GSD artifacts: quick task `260914-pbr` planning and summary.

## Quality gates

- `npm run typecheck`: clean
- `npm run lint`: clean
- `npm run fallow`: clean
- `npm run format:check`: clean
- `npm run test:corresponding`: clean
- `node scripts/test-coverage-direct.mjs --base upstream/main`: clean (0 unpinned shortfalls)
- `trufflehog`: clean (0 secrets)

### CODE REVIEW

**Comment by rakesh-vs** [comment]
@acolomba what do you think of this change? It's a great UX improvement IMO.

**Comment by acolomba** [comment]
@rakesh-vs , yes, an interactive ui would be great. i had an agent review this when it was first submitted, and it found a number of gaps. (now i wish i'd posted them, but i'm still figuring out the etiquette for agentic collaboration.)

you may try to have your own agent iterate on it, and ideally use gsd so that the pr is mergeable. otherwise i'll take care of it once i have agent time available. (right now i have mine busy on quality refactors.)

whichever way you want to do it. thanks!

**Comment by rakesh-vs** [comment]
Thanks @acolomba ! Rebased on main and resolved the gaps with full GSD artifacts:
- 100% direct coverage across `plugin-browser.ts` and `browse.ts`.
- Non-TUI fallback to `list`.
- Added CHANGELOG entry, docs, and GSD planning metadata (`.planning/quick/`).
- Split into three commits: feature + tests, docs, and GSD task.

All gates pass (`npm run check`, fallow, direct coverage).
Let me know if anything is missing!

File: extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+/**
+ * Build the arg string the typed subcommand handler expects. Every action
+ * targets a single `<plugin>@<marketplace>` in the marketplace's scope,
+ * with optional `--local` targeting `claude-plugins.local.json`.
+ */
+function actionArgs(result: PickerResult): string {
+  const local = result.local === true ? " --local" : "";
+  return `${result.plugin}@${result.marketplace} --scope ${result.scope}${local}`;
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts:97]: **Wrong scope dispatched for uninstall / disable / enable / info.**

`actionArgs` always emits `--scope ${result.scope}`, and that scope is the *marketplace record's* scope (`MarketplaceEntry.scope`, seeded from `loadVisibleMarketplaces`). But the rows this picker is built from carry their own: `ScopeBearingListStatus` (`installed` / `disabled` / `upgradable` / `partially-installed` / `partially-upgradable`) declares `scope?: Scope` in `shared/notification-types.ts`.

The sibling handler already solves this. `edge/handlers/tools.ts:367`:

```ts
function pluginScopeOrFallback(
  p: PluginNotificationMessage,
  marketplaceScope: "user" | "project",
): "user" | "project" {
  return isScopeBearingListRow(p) ? (p.scope ?? marketplaceScope) : marketplaceScope;
}
```

`browse.ts` otherwise mirrors `tools.ts` closely -- same `loadVisibleMarketplaces` + `loadPluginListPayload` pair, same `N plugin(s) -- <source>` row shape -- but this one call was dropped.

Failure case: a user-scope marketplace whose plugin was installed into project scope. The picker dispatches `uninstall x@mp --scope user` and misses the install record.

It's worse than merely guessing, because `UninstallPluginOptions.scope` is *optional* (`orchestrators/plugin/uninstall.ts:135`) and the handler forwards it only when present (`edge/handlers/plugin/uninstall.ts:40`) -- so omitting `--scope` would search both scopes and find it. Forcing an explicit wrong scope is strictly worse than forcing none.

Suggested fix: resolve the scope at projection time in `loadPlugins` via `pluginScopeOrFallback`, carry it on `PluginEntry`, and use it for the four non-install actions. `install` correctly keeps whatever the user picks on the scope screen.

File: extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+        return version !== undefined
+          ? { name: p.name, status: p.status, version }
+          : { name: p.name, status: p.status };
+      });
+      this.pluginsByName.clear();
+      for (const entry of this.plugins) {
+        this.pluginsByName.set(entry.name, entry);
+      }
+
+      this.buildPluginsList(mp);
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts:354]: **Stale-load race: the resolving loader clobbers whatever screen the user is on now.**

`showPlugins` paints the `loading…` frame with `this.list = null`, then fires `void this.loadPlugins(mp)`. While that's in flight, `handleInput`'s `this.list === null` branch (line 455) sends Esc to `showMarketplaces()`. When the promise resolves, this line calls `buildPluginsList(mp)` unconditionally and repaints over the marketplaces screen the user just navigated back to.

Repro: open a marketplace whose loader is slow (cold git-source clone probe), press Esc, watch the plugin list appear over the marketplaces screen a moment later.

Second instance of the same defect: open marketplace A, Esc, open marketplace B. Two loads are in flight, whichever resolves last wins, so A's plugins can land under B's title. The `catch` arm has it too -- a failed load for A repaints "Failed to load plugins" over B's screen.

Suggested fix: a monotonic epoch captured before the await and compared after, bumped by every screen transition that abandons a load:

```ts
private loadEpoch = 0;

private async loadPlugins(mp: MarketplaceEntry): Promise<void> {
  const epoch = ++this.loadEpoch;
  try {
    const rows = await this.pluginLoader(mp);
    if (epoch !== this.loadEpoch) { return; }
    // ...project + buildPluginsList
  } catch {
    if (epoch !== this.loadEpoch) { return; }
    // ...failure frame
  }

  this.tui.requestRender();
}
```

with `showMarketplaces()` bumping `loadEpoch` so an Esc invalidates the in-flight load.

File: extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+
+  private showMarketplaces(): void {
+    const items: SelectItem[] = this.marketplaces.map((m) => ({
+      value: m.name,
+      label: `${m.name} [${m.scope}]`,
+      description: `${m.pluginCount.toString()} plugin(s) -- ${m.source}`,
+    }));
+    const list = new SelectList(items, Math.min(items.length, MAX_VISIBLE), this.listTheme());
+    list.onSelect = (item) => {
+      const mp = this.marketplaces.find((m) => m.name === item.value);
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts]: **Same-name marketplaces across scopes collide (1 of 2).**

`SelectItem.value` is `m.name` (line 274) and this resolves the pick with `find(m => m.name === item.value)`. Names are not unique across entries: `loadVisibleMarketplaces` (`orchestrators/marketplace/shared.ts:577`) walks `["project", "user"]` and pushes **every** record from both with no dedup -- unlike `list-flow.ts:634-640`, which explicitly folds a project-scope record that is a clone of the user-scope one (`isOrphanMarketplaceClone`) precisely so the surface doesn't show it twice.

So a marketplace present in both scopes renders two rows (`foo [project]`, `foo [user]`) and selecting the second opens the first.

Suggested fix: put the scope in the item value -- `value: `${m.scope}:${m.name}`` with the label unchanged -- or key a `Map` on the composite. The same composite key then fixes the loader side (see the comment on `makePluginLoader`).

File: extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+ * display entries.
+ */
+function makePluginLoader(
+  pi: ExtensionAPI,
+  ctx: ExtensionCommandContext,
+  cwd: string,
+): (marketplace: MarketplaceEntry) => Promise<readonly PluginNotificationMessage[]> {
+  return async (mp): Promise<readonly PluginNotificationMessage[]> => {
+    const payload = await loadPluginListPayload({ ctx, pi, cwd, marketplace: mp.name });
+    const block = payload.find((b) => b.name === mp.name);
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts]: **Same-name marketplaces across scopes collide (2 of 2).**

`payload.find((b) => b.name === mp.name)` takes the first block with a matching name, but `loadPluginListPayload` can emit two blocks for one name -- one per scope -- whenever the project-scope record is *not* an orphan clone of the user-scope one (`list-flow.ts:627-641` skips only the clone case). The `marketplace: mp.name` filter narrows by name, not by scope, so it doesn't collapse them either.

Result: for a genuinely distinct same-name pair, the second block's plugins are unreachable through the picker, and both marketplace rows show the first block's plugins.

Pairs with the `showMarketplaces` lookup in `plugin-browser.ts` -- one composite name+scope key fixes both sides.

File: extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+      return;
+    }
+
+    // Eager marketplace load. The plugins for a marketplace load lazily
+    // inside the browser when the marketplace is opened.
+    let marketplaces;
+    try {
+      marketplaces = await loadMarketplaceEntries(ctx.cwd);
+    } catch (err) {
+      notifyDiagnostic(ctx, "/claude:plugin browse", [
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts]: **`notifyDiagnostic` is the cascade seam, not a general edge-handler surface.**

Its doc comment scopes it deliberately (`shared/notification-dispatch.ts:104-122`): it is the "post-cascade hygiene warnings out-of-band notification seam", its header is documented as counting the per-warning lines, and it is called out as the seam for "orchestrated-mode (cascade) callers". The only two other call sites are `orchestrators/plugin/shared.ts:1424` and `orchestrators/reconcile/apply.ts:960`. Here it gets a bare command name as the header and free prose as the body, bypassing `shared/notification-grammar.ts` -- the closed-vocabulary renderer everything else routes through.

Two concrete consequences:

1. **Severity is wrong on the failure path.** The seam hardcodes `"warning"` (line 132). `Failed to load marketplaces: …` is a command that did *not* carry out, which is `error`, not `warning`.
2. **"No marketplaces" already has a house rendering** -- the `(no marketplaces)` sentinel, emitted through the structured message path at info severity. See `orchestrators/marketplace/list.ts:100`, `autoupdate.ts:552`, and `import/execute.ts:613` ("`(no marketplaces)` with no severity argument -- info").

Suggested fix: emit the empty case through the structured notification path so it renders the same sentinel every other surface does, and raise the load failure to `error`.

File: extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+      return "[remote]";
+    case "partially-available":
+    case "partially-installed":
+    case "partially-upgradable":
+      return "[partial]";
+    case "unavailable":
+      return "[unavailable]";
+    case "disabled":
+      return "[disabled]";
+    default:
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts]: **`default:` on a closed-set union defeats the exhaustiveness check.**

`statusTag` (this switch), `statusDescription` (line 137), and `availableActions` (line 167) each terminate in a `default:` over `PluginNotificationMessage["status"]`. That is a closed set, but the `default:` means adding a list-surface status compiles clean at all three sites while the picker silently renders an empty tag, an empty description, and offers only `info`.

The house pattern is the opposite: `assertNever` appears in 14 modules for exactly this, and `tools.ts:442`'s `pluginVersion` is deliberately a value-returning switch with **no** default arm (its header cites D-116-14 for the choice).

Suggested fix: enumerate the transition arms explicitly -- they're already named in the comment on line 131 -- and `assertNever` the tail, so a future status becomes a type error at all three sites instead of three silent blanks.

File: extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+  ) => Promise<readonly PluginNotificationMessage[]>;
+  readonly onSelect: (result: PickerResult) => void;
+  readonly onCancel: () => void;
+}
+
+const MAX_VISIBLE = 12;
+
+// ─── display helpers ──────────────────────────────────────────────────────
+
+function pluginVersion(p: PluginNotificationMessage): string | undefined {
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts]: **Second `pluginVersion`, and this one is structurally typed.**

`edge/handlers/tools.ts:442` already has a `pluginVersion(p): string | undefined` over the same union, written as an exhaustive switch on `p.status`. This one duck-types with `"version" in p && typeof p.version === "string"`.

It sits under fallow's `duplicates.threshold: 3` so the gate stays green, but it is two implementations of one concept with opposite failure modes: the switch stops compiling when the union grows, the `in` check silently keeps returning `undefined`.

Same root cause as the `default:`-arm comment above -- worth hoisting one typed implementation somewhere both surfaces can reach.

File: extensions/pi-claude-marketplace/edge/router.ts [hunk 8]
```diff
@@ -102,7 +106,8 @@ export const TOP_LEVEL_USAGE =
   "  enable <plugin>@<marketplace> [--scope user|project] [--local]\n" +
   "  disable <plugin>@<marketplace> [--scope user|project] [--local]\n" +
   "  import [--scope user|project]\n" +
-  "  marketplace <add|remove|rm|list|ls|info|update|autoupdate|noautoupdate> ...";
+  "  marketplace <add|remove|rm|list|ls|info|update|autoupdate|noautoupdate> ...\n" +
+  "  browse                                             interactive marketplaces -> plugins -> action picker";
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/router.ts]: **The synopsis line above still omits `browse`, and the ordering disagrees with the vocabulary array.**

Line 96 -- the canonical one-line vocabulary -- is unchanged:

```
Usage: /claude:plugin <bootstrap|install|uninstall|update|fetch|reinstall|list|ls|info|pending|enable|disable|import|marketplace> ...
```

`browse` gets a body line but never appears in the synopsis, so the usage block now contradicts itself.

Also `browse` is last here but sits before `marketplace` in `TOP_LEVEL_SUBCOMMANDS` (line 75). That array is the completion surface in declaration order and `tests/edge/completions/provider.test.ts` TC-1 pins it, so the two lists are now deliberately different orders for the same vocabulary. Worth making them agree.

Minor: this is the only body line besides `bootstrap` carrying an aligned trailing description; the rest are bare signatures.

**Review by acolomba — COMMENTED** [review]
Reviewed at `d33f55d`. All checks green -- `npm run check`, fallow, direct coverage, pre-commit, integration, pinned e2e.

Good shape overall, and the design instincts are right. `pluginLoader` is genuinely injected, so `PluginBrowser` carries no `ctx`/`pi`/orchestrator coupling. `DynamicBorder` and `Theme` go through the `platform/pi-api.ts` chokepoint. Dispatching into the existing typed handlers keeps arg-parsing and notify surfaces identical rather than forking them. The tests are hermetic -- real tmpdirs, real state seeding, real `SelectList` keystrokes, no module mocking. And `demos/` correctly stays out of the npm tarball.

I checked the peer floor too, since `ctx.mode`, `ui.custom`, `DynamicBorder`, and `Theme` are all new call surfaces for this repo: `ctx.mode` landed in pi **0.78.1**, below the declared `>=0.80.5` floor, and the other three predate it. No NFR-11 problem.

Three correctness bugs want fixing before this lands, all commented inline:

1. **Wrong scope** dispatched for uninstall/disable/enable/info. `tools.ts` already has the helper this drops (`pluginScopeOrFallback`), and forcing a wrong `--scope` is worse than omitting it, since the orchestrator's `scope` is optional.
2. **Stale-load race** in `loadPlugins` -- an Esc, or a second marketplace opened during the load, gets repainted over by the resolving promise.
3. **Same-name marketplaces across scopes collide**, in both the picker lookup and the payload block lookup. `list-flow.ts` folds this case deliberately; `loadVisibleMarketplaces` does not.

Then three convention items -- the `notifyDiagnostic` seam and its hardcoded severity, `default:` arms defeating exhaustiveness on a closed status set, and a second `pluginVersion` -- plus smaller notes on the usage block and the discarded `browse` args.

One thing I'd like added rather than just flagged: bare `/claude:plugin` used to be how you read the usage block, and now nothing reaches it except a typo. Routing bare input to the picker is correct, but please add a `help` subcommand to take over that job -- specifics in the comment on the `head === ""` branch.

Docs side is fine as-is: PRD AP-3 amended to match the new routing, both READMEs updated, CHANGELOG entry reads well.

File: extensions/pi-claude-marketplace/edge/router.ts [hunk 9]
```diff
@@ -141,8 +146,10 @@ export async function routeClaudePlugin(
   const [head, rest] = peelToken(args);
 
   if (head === "") {
-    notifyUsageError(ctx, { message: "Usage error.", usage: TOP_LEVEL_USAGE });
-    return;
+    // Bare `/claude:plugin` opens the interactive browser (TUI) or falls
+    // back to the inventory list (non-TUI), matching Claude Code's
+    // `/plugin` UX. `rest` is "" here (peelToken of empty input).
+    return handlers.browse(rest, ctx);
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/router.ts:156]: **Bare `/claude:plugin` was how you read the usage block -- please add a `help` subcommand to take that over.**

Before this change `head === ""` emitted `TOP_LEVEL_USAGE`. Now the only path that reaches it is typing an *invalid* subcommand: the user has to make a mistake to read the docs, and in non-TUI modes there is no way to ask for them at all, since bare input runs `list`.

Routing bare input to the picker is the right call for `/plugin` parity and the PRD's AP-3 was amended to match -- no objection to the behavior itself. But the usage surface shouldn't be collateral damage. Suggestion:

- add `"help"` to `TOP_LEVEL_SUBCOMMANDS` and to the `Usage:` synopsis on line 96,
- add `case "help": return handlers.help(rest, ctx);` to the switch below, rendering `TOP_LEVEL_USAGE` at **info** severity -- not via `notifyUsageError`, which prefixes `Usage error.` and stamps `error`; nothing failed here,
- let `help marketplace` (or `marketplace help`) render `MARKETPLACE_USAGE`, which has the same problem -- `routeMarketplace` also surfaces it only on bad input,
- add the `help` row to `TOP_LEVEL_DISPATCH` in `tests/edge/router.test.ts` and to the TC-1 vocabulary list in `tests/edge/completions/provider.test.ts`.

That restores discoverability in every mode, and gives the new `browse` usage line somewhere honest to sit next to.

File: extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+ * Factory: returns the async handler closed over `pi` (needed by the
+ * plugin loader's `loadPluginListPayload` soft-dep probe) and the action
+ * handlers.
+ */
+export function makeBrowseHandler(
+  pi: ExtensionAPI,
+  handlers: BrowseActionHandlers,
+): (args: string, ctx: ExtensionCommandContext) => Promise<void> {
+  return async (_args, ctx): Promise<void> => {
+    // Non-TUI modes have no terminal for `ctx.ui.custom`; fall back to the
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts:138]: **`_args` is discarded, so `browse` silently accepts anything.**

`/claude:plugin browse --scope user`, `/claude:plugin browse foo`, and `/claude:plugin browse` all do exactly the same thing.

Every other subcommand rejects unknown input -- `info.ts:41-45` is the pattern -- and `flag-catalog.ts` has no `browse` entry, so the accepted set here is legitimately empty. Not a correctness bug, but a usage-error surface the rest of the router maintains consistently; a short "browse takes no arguments" error on non-empty input would match.

File: extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts (hunk 1 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts:97]: Resolved scope on `PluginEntry` via `pluginScopeOrFallback` and forwarded it to uninstall, info, enable, and disable.

File: extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts (hunk 2 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts:354]: Added monotonic `loadEpoch` checked before and after async operations, incremented on screen transitions.

File: extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts (hunk 3 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts]: Keyed marketplace select items and loader lookups by `${scope}:${name}`.

File: extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts (hunk 4 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts]: Matched both `b.name === mp.name && b.scope === mp.scope` in `makePluginLoader`.

File: extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts (hunk 5 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts]: Empty marketplace list now notifies through the structured path to render `(no marketplaces)`, and load failure emits `error` severity.

File: extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts (hunk 6 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts]: Enumerated transition arms explicitly and added `default: return assertNever(status)` across all three switches.

File: extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts (hunk 7 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/browser/plugin-browser.ts]: Hoisted typed `pluginVersion` to `shared/notification-types.ts` and removed the duplicate in `tools.ts`.

File: extensions/pi-claude-marketplace/edge/router.ts (hunk 8 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/router.ts]: Added `browse` and `help` to the synopsis line and aligned their ordering with `TOP_LEVEL_SUBCOMMANDS`.

File: extensions/pi-claude-marketplace/edge/router.ts (hunk 9 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/router.ts:156]: Added `/claude:plugin help` and `marketplace help` rendering usage at info severity, with completions and router dispatch tests.

File: extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts (hunk 10 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts:138]: Added usage error check rejecting non-empty arguments to `browse`.

**Comment by rakesh-vs** [comment]
@acolomba Thanks for the detailed review! Fixed the scope resolution, stale-load races, same-name collisions, and added `/claude:plugin help` along with the convention cleanups. All checks passing with 100% direct coverage.

File: extensions/pi-claude-marketplace/edge/router.ts [hunk 11]
```diff
@@ -211,6 +225,8 @@ async function routeMarketplace(
       return handlers.marketplaceAutoupdate(rest, ctx);
     case "noautoupdate":
       return handlers.marketplaceNoautoupdate(rest, ctx);
+    case "help":
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/router.ts:229]: **`marketplace help` is routable but invisible to both the vocabulary and the census.**

This arm was added to the switch, but `"help"` never went into `MARKETPLACE_SUBCOMMANDS` (line 85) or into `MARKETPLACE_DISPATCH` in `tests/edge/router.test.ts`. Two consequences:

1. **No completion.** `MARKETPLACE_SUBCOMMANDS` is what `completions/provider.ts:126` filters to build marketplace-subcommand suggestions, so `marketplace help` never appears while every sibling does.
2. **The census test silently stopped censusing.** `router.test.ts:167` is titled *"accepts exactly the marketplace names its dispatch rows serve"* and asserts `[...MARKETPLACE_SUBCOMMANDS]` deep-equals `MARKETPLACE_DISPATCH.map(r => r.subcommand)`. Both sides are unchanged, so it passes -- while the switch it is supposed to describe grew a third arm that appears in neither list. The standalone test added at `router.test.ts:178` proves the routing works, but it does not restore the totality claim the census makes.

Same applies at the top level in the mirror image: `"help"` *was* added to `TOP_LEVEL_SUBCOMMANDS`, so that side is consistent. It's only the marketplace-nested arm that drifted.

Suggested fix: add `"help"` to `MARKETPLACE_SUBCOMMANDS` and a matching row to `MARKETPLACE_DISPATCH`. The dispatch row needs the special arg (`handlers.help("marketplace", ctx)` rather than `rest`), so the generic row runner will need an optional expected-args field -- which is presumably why the standalone test was written instead. Worth doing anyway so the census stays total.

Not your problem to solve in this PR, for context: the underlying reason one subcommand has to be declared in four independent places (`SubcommandHandlers`, the vocabulary array, `TOP_LEVEL_USAGE`, and the switch) is that `peelToken` returns `string`, so `switch (head)` is a switch over an open type and can never be exhaustive. Tightening that is being tracked separately -- please just add the two list entries here.

File: extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+      await handlers.list("", ctx);
+      return;
+    }
+
+    // Eager marketplace load. The plugins for a marketplace load lazily
+    // inside the browser when the marketplace is opened.
+    let marketplaces;
+    try {
+      marketplaces = await loadMarketplaceEntries(ctx.cwd);
+    } catch {
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts]: **The load-failure cause is now discarded entirely.**

This `catch` binds nothing, and the payload carries no `reasons`, so a failure renders as a `(browse)` row with no explanation at all. The prose this replaced at least said `Failed to load marketplaces: ${errorMessage(err)}`. Moving off the old seam was right; losing the cause with it was not.

The reason is a modeling detail rather than an oversight: `MpFailed` (`shared/notification-types.ts:466`) declares `reasons?` but has no `cause?: Error`, unlike `PluginFailedMessage` (`:284`), so the arm chosen here has nowhere to put the error.

`reinstall-flow.ts:636-655` is the house shape for exactly this case -- a synthetic-name failure with a real cause behind it:

```ts
const reasons: readonly ContentReason[] =
  typedReasons ?? narrowReasons([composeErrorWithCauseChain(err)]);
const failedRow: PluginFailedMessage = {
  status: "failed",
  name: "(reinstall)",
  reasons,
  cause: causeErr,
  severity: "error",
  needsReload: false,
};
notifyWithContext(ctx, pi, REINSTALL_CONTEXT,
  [{ name: targetingMp, scope: targetingScope, plugins: [failedRow] }], undefined);
```

The failure rides as a `PluginFailedMessage` *nested under* the synthetic marketplace row, rather than the marketplace row itself being `status: "failed"` with `plugins: []`. That arm can carry both `reasons` and `cause`.

For what it's worth I checked the two things here that looked wrong at first glance and they're both fine: the parenthesized `(browse)` name matches the `(list)` / `(reinstall)` convention, and the hardcoded `"user"` scope matches `reinstall-flow.ts:639`'s `opts.scope ?? "user"` fallback. Only the dropped cause needs addressing.

File: extensions/pi-claude-marketplace/shared/notification-types.ts [hunk 13]
```diff
@@ -399,6 +399,47 @@ export function isScopeBearingListRow(
...
+  p: PluginNotificationMessage,
+  marketplaceScope: Scope,
+): Scope {
+  return isScopeBearingListRow(p) ? (p.scope ?? marketplaceScope) : marketplaceScope;
+}
+
+/**
+ * Read `p.version` off a plugin notification row.
+ */
+export function pluginVersion(p: PluginNotificationMessage): string | undefined {
```
Comment (acolomba) [inline extensions/pi-claude-marketplace/shared/notification-types.ts:433]: **The hoist dropped the decision IDs that explained these two functions.**

Both helpers moved up from `tools.ts`, and both lost their rationale on the way.

`pluginVersion` previously carried **D-15-04** (every list-surface variant shares the same optional `version?` slot, so every arm returns the same field and the switch computes nothing) and **D-116-14**, which is the load-bearing one:

> the switch stays anyway, and must not be collapsed into a single expression. Its job is the missing-arm gate -- `noImplicitReturns` makes the end of this function reachable the moment a list-surface status goes unnamed

Without that note, the next reader sees a switch where every arm returns `p.version` and reasonably "simplifies" it, silently removing the gate.

`pluginScopeOrFallback` (line 406) lost **SNM-11** and **D-13-18** the same way -- specifically the explanation that `available` / `unavailable` omit `scope` by construction, and that the other variants carry it only when the install scope differs from the marketplace block's scope under the orphan-fold rule. That's the whole reason the fallback parameter exists.

`.claude/rules/typescript-comments.md` lists decision and requirement IDs under "Allowed (and encouraged) as traceability anchors", so carrying them across a move is the house expectation. Mechanical to restore -- both texts are in the deleted hunks of `tools.ts` in this diff.

One substantive note while you're here: the union widened on the move, and `case "updated": return p.to` is a new semantic that neither call site can currently reach. Fine as written -- just flagging it as a deliberate-looking choice that no test exercises, which is exactly the kind of thing D-116-14's gate exists to force a decision on.

**Review by acolomba — COMMENTED** [review]
Re-reviewed at `ea375f9b`, all eight checks green. **Every one of the ten comments from the last pass is addressed**, and most of them properly rather than superficially. Specifically verified rather than assumed:

- **Scope** -- `PluginEntry` gained `scope`, resolved through `pluginScopeOrFallback(p, mp.scope)` at projection time, and `onSelect` sends `plugin.scope`. The action-screen description line was updated to match.
- **Load race** -- `loadEpoch` is bumped in both `loadPlugins` and `showMarketplaces` and checked after the await in *both* the try and the catch. I walked the three interleavings (Esc-during-load, A-then-B, failure-during-B); all correct now.
- **Name collision** -- composite `${scope}:${name}` key on the picker side, `b.name === mp.name && b.scope === mp.scope` on the loader side. `MpCommon.scope` is required, so the block predicate is sound.
- **Exhaustiveness** -- all three switches enumerate the transition arms and `assertNever` the tail.
- **Duplicate `pluginVersion`** -- hoisted to `shared/`, `tools.ts` down 62 lines, and the new copy has no `default:` arm so the `noImplicitReturns` gate survives.
- **Usage block** -- synopsis, body order, and `TOP_LEVEL_SUBCOMMANDS` now agree.
- **`browse` args** and **`help`** -- both built as discussed, and the empty-marketplace case now renders the `(no marketplaces)` sentinel through `notify()`.

Three follow-ups, all inline and all small:

1. **`marketplace help`** is routable but absent from `MARKETPLACE_SUBCOMMANDS` and `MARKETPLACE_DISPATCH`, so it gets no completion and the census test's totality claim is no longer true.
2. **The marketplace-load failure drops its cause.** Right call to leave the old seam; the replacement arm just can't carry an error. `reinstall-flow.ts` shows the shape that can.
3. **Decision IDs (D-15-04, D-116-14, SNM-11, D-13-18) were lost** in the helper hoist. Mechanical to restore from the deleted `tools.ts` hunks in this same diff.

None of these are structural -- this is close.

One process note for next time: the fixes came in as a force-push rather than new commits. The repo's contributing rules ask for merges over history rewrites; low practical impact here since this branch squash-merges, but it does make re-reviewing harder because the old review comments no longer anchor to reachable blobs.

**Comment by acolomba** [comment]
@rakesh-vs , leaving another few comments. generally i run pr-review-toolkit (from the claude marketplace) in a loop.

there's a babysit-pr skill in the repo that does that on its own if you want to try that, but no problem otherwise. i need to do a follow-up anyway to address a design gap that this pr surfaced about commands and completion, which doesn't need to be in this pr.

thanks!

File: extensions/pi-claude-marketplace/edge/router.ts (hunk 11 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/router.ts:229]: Added "help" to `MARKETPLACE_SUBCOMMANDS`, updated completion provider tests, and added the census row with `expectedArgs` in `MARKETPLACE_DISPATCH`.

File: extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts (hunk 12 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/edge/handlers/plugin/browse.ts]: Switched to the `PluginFailedMessage` nested row pattern from `reinstall-flow.ts` so the error cause is retained and rendered.

File: extensions/pi-claude-marketplace/shared/notification-types.ts (hunk 13 above)
Comment (rakesh-vs) [reply extensions/pi-claude-marketplace/shared/notification-types.ts:433]: Restored the full comments and decision IDs (D-15-04, D-116-14, SNM-11, D-13-18).

**Comment by rakesh-vs** [comment]
Thanks for the catch. Pushed as a clean new commit this time.

Addressed all three follow-ups:
- Added `help` to `MARKETPLACE_SUBCOMMANDS` and `MARKETPLACE_DISPATCH` census row with `expectedArgs`.
- Preserved load failure cause using a nested `PluginFailedMessage` under the `(browse)` marketplace row.
- Restored D-15-04, D-116-14, SNM-11, and D-13-18 decision IDs on the hoisted helpers.

**Review by acolomba — APPROVED** [review]
@rakesh-vs thank you!
